utils/python/CIME/macros.py

```diff
@@ -0,0 +1,1007 @@
+"""
+Classes used to build the CIME Macros file.
+
+The main "public" class here is MacroMaker. Other classes are mainly for
+internal use, though the various "MacroWriter" classes might be useful
+elsewhere one day.
```

Comment:
I can definitely do more in the way of design documentation. I'm also considering putting the "Writer" classes in a different module, if that would help navigate through this module.